### PR TITLE
gh-94603: micro optimize list.pop

### PR DESCRIPTION
**Pull request by eendebakpt**
This PR optmizes `list.pop`

Benchmark results
```
x=[1,2]; x.pop(0): Mean +- std dev: [m2] 81.8 ns +- 2.5 ns -> [pr_special] 68.9 ns +- 2.0 ns: 1.19x faster
x=[1,2]; x.pop(1): Mean +- std dev: [m2] 65.4 ns +- 2.1 ns -> [pr_special] 64.4 ns +- 0.6 ns: 1.02x faster
x=[]; x.insert(0,1); x.pop(0): Mean +- std dev: [m2] 111 ns +- 3 ns -> [pr_special] 80.2 ns +- 1.1 ns: 1.38x faster
list(0): insert(0, None) + pop(0): Mean +- std dev: [m2] 78.4 ns +- 0.8 ns -> [pr_special] 62.1 ns +- 1.3 ns: 1.26x faster
list(1): insert(0, None) + pop(1): Mean +- std dev: [m2] 59.7 ns +- 0.4 ns -> [pr_special] 57.7 ns +- 0.9 ns: 1.04x faster
list(2): insert(0, None) + pop(1): Mean +- std dev: [m2] 66.9 ns +- 0.6 ns -> [pr_special] 55.5 ns +- 1.8 ns: 1.20x faster
list(10): insert(0, None) + pop(1): Mean +- std dev: [m2] 71.9 ns +- 0.6 ns -> [pr_special] 62.4 ns +- 2.8 ns: 1.15x faster
list(100): insert(0, None) + pop(1): Mean +- std dev: [m2] 108 ns +- 4 ns -> [pr_special] 93.5 ns +- 3.3 ns: 1.16x faster
list(10000): insert(0, None) + pop(1): Mean +- std dev: [m2] 5.41 us +- 0.02 us -> [pr_special] 5.37 us +- 0.02 us: 1.01x faster

Benchmark hidden because not significant (2): x=[1]; x.pop(0), sortedcontainer

Geometric mean: 1.12x faster
```
<details>
<summary>Benchmark code</summary>

```
import pyperf
import sortedcontainers


runner = pyperf.Runner()

runner.timeit(name=f"x=[1]; x.pop(0)",stmt='x=[1]; x.pop(0)')
runner.timeit(name=f"x=[1,2]; x.pop(0)", stmt='x=[1,2]; x.pop(0)')
runner.timeit(name=f"x=[1,2]; x.pop(1)",stmt='x=[1,2]; x.pop(1)')
runner.timeit(name=f"x=[]; x.insert(0,1); x.pop(0)",stmt='x=[]; x.insert(0,1); x.pop(0)')
runner.timeit(name=f"list(0): insert(0, None) + pop(0)", stmt="lst.insert(0, None); lst.pop(0)", setup='lst=[]')

for n in [ 1, 2, 10, 100, 10_000]:
    runner.timeit(name=f"list({n}): insert(0, None) + pop(1)",
	      stmt="lst.insert(0, None); lst.pop(1)", setup=f'import random; lst=[random.random() for ii in range({n})]')

setup ="import random; import sortedcontainers; sc=sortedcontainers.SortedList(); x=[random.random() for n in range(1_00_000)]"

runner.timeit(name=f"sortedcontainer", stmt="[sc.add(v) for v in x]", setup=setup)
```
</details>

* Issue: gh-94603

### CODE REVIEW

**Comment by corona10** [comment]
@serhiy-storchaka is the expert who can decide these kinds of micro-optimization is worth applying :)

File: Objects/listobject.c [hunk 1]
```diff
@@ -1040,21 +1040,26 @@ list_pop_impl(PyListObject *self, Py_ssize_t index)
...
-        else
-            return NULL;
     }
-    Py_INCREF(v);
-    status = list_ass_slice(self, index, index+1, (PyObject *)NULL);
-    if (status < 0) {
-        Py_DECREF(v);
+    if (status >= 0)
+        return v; /* and v now owns the reference the list had */
+    else {
```
Comment (corona10) [inline Objects/listobject.c]:
```suggestion
    if (status >= 0) {
        return v; // and v now owns the reference the list had 
    }
    else {
```

nit PEP7

**Comment by serhiy-storchaka** [comment]
I am a bit surprised to see a significant difference in that microbenchmarks. The method call adds an overhead, and since this method is not idempotent, you need to add even more expensive code to create a list or adding an item in the existing list list at every iteration, so I expected the difference be rather unnoticeable. But I repeated microbenchmarking myself and confirm the results. It looks interesting.

Now, we need to analyze, why we get such significant speedup, can the optimization be generalized to other methods, can we get the same results with simpler changes?

File: Objects/listobject.c [hunk 2]
```diff
@@ -1040,21 +1040,27 @@ list_pop_impl(PyListObject *self, Py_ssize_t index)
...
+
+    PyObject **items = self->ob_item;
+    v = items[index];
+    if(Py_SIZE(self)-1==0) {
+        Py_INCREF(v);
+        status = _list_clear(self);
+    }
+    else {
+        if (Py_SIZE(self)-1-index)
+            memmove(&items[index], &items[index+1], (Py_SIZE(self)-1-index)* sizeof(PyObject *));
```
Comment (corona10) [inline Objects/listobject.c]: ditto
```suggestion
        if (Py_SIZE(self)-1-index) {
            memmove(&items[index], &items[index+1], (Py_SIZE(self)-1-index)* sizeof(PyObject *));
        }
```

**Comment by eendebakpt** [comment]
@serhiy-storchaka Here are some more details

* The `test/test_list.py` is not part of the PGO cycle. The `list.pop` is called multiple times though, so it is part of the profiling so I did not change that
* The specialization for popping the last element of a list is required to achieve the same performance as current `main`. Benchmarks on `runner.timeit(name=f"pop last item",stmt='x=[1,2,3]; x.pop(-1)')`:

Main
```
pop last item: Mean +- std dev: 68.2 ns +- 0.8 ns
```
PR
```
pop last item: Mean +- std dev: 65.0 ns +- 1.1 ns
```
PR: No specialization for last element
```
pop last item: Mean +- std dev: 73.6 ns +- 0.5 ns
```

* The specialization for empty lists is required as well. Benchmark `runner.timeit(name=f"pop to empty",stmt='x=[1,]; x.pop(0)')`:

Main
```
pop to empty: Mean +- std dev: 64.0 ns +- 0.4 ns
```
PR
```
pop to empty: Mean +- std dev: 61.5 ns +- 0.9 ns
```
PR: No specialization for last element
```
pop to empty: Mean +- std dev: 64.5 ns +- 0.4 ns
```

(differences are small though)

* The reason for the speedup might be that we avoid an extra call to `memcopy` in `list_ass_slice` (e.g. we replace `memcpy(recycle, &item[ilow], s);` with a direct assignment)

**Comment by eendebakpt** [comment]
Added similar optimizations to `list_ass_slice` does not seem to have the same performance benefits. Not sure why that is

https://github.com/python/cpython/compare/main...eendebakpt:list_assignment_optimization

results in:
```
x=[1,2]; x.pop(0): Mean +- std dev: [m5] 80.8 ns +- 1.1 ns -> [pr_slice3] 79.3 ns +- 3.2 ns: 1.02x faster
x=[1,2]; x.pop(1): Mean +- std dev: [m5] 65.0 ns +- 0.4 ns -> [pr_slice3] 66.7 ns +- 2.1 ns: 1.03x slower
list(0): insert(0, None) + pop(0): Mean +- std dev: [m5] 79.1 ns +- 0.6 ns -> [pr_slice3] 83.8 ns +- 0.5 ns: 1.06x slower
list(50): x.append(1); x.pop(-2): Mean +- std dev: [m5] 48.1 ns +- 1.4 ns -> [pr_slice3] 45.6 ns +- 0.3 ns: 1.06x faster
list(1): insert(0, None) + pop(1): Mean +- std dev: [m5] 57.6 ns +- 1.0 ns -> [pr_slice3] 61.7 ns +- 0.3 ns: 1.07x slower
list(2): insert(0, None) + pop(1): Mean +- std dev: [m5] 64.8 ns +- 0.5 ns -> [pr_slice3] 68.0 ns +- 2.2 ns: 1.05x slower
list(10): insert(0, None) + pop(1): Mean +- std dev: [m5] 69.1 ns +- 1.0 ns -> [pr_slice3] 73.3 ns +- 1.8 ns: 1.06x slower
list(100): insert(0, None) + pop(1): Mean +- std dev: [m5] 119 ns +- 2 ns -> [pr_slice3] 108 ns +- 4 ns: 1.11x faster
list(10000): insert(0, None) + pop(1): Mean +- std dev: [m5] 5.82 us +- 0.03 us -> [pr_slice3] 5.40 us +- 0.16 us: 1.08x faster
sortedcontainer: Mean +- std dev: [m5] 105 ms +- 1 ms -> [pr_slice3] 110 ms +- 4 ms: 1.05x slower

Benchmark hidden because not significant (2): x=[1]; x.pop(0), x=[]; x.insert(0,1); x.pop(0)

Geometric mean: 1.00x slower
```

**Comment by eendebakpt** [comment]
> Added similar optimizations to `list_ass_slice` does not seem to have the same performance benefits. Not sure why that is
A small variant https://github.com/python/cpython/compare/main...eendebakpt:list_assignment_optimization2 also gives performance similar to `main`. Avoiding a single memcpy does not fully explain the performance gain.

The `list_ass_slice` does a lot more work than is required for the case of a simple pop. When calling from `list_pop_impl` there are 7 redundant if statements. E.g. if-statements that are required for the implementation of `list_ass_slice`, but that would not be required if the method is only called from `list_pop_impl`.
The additional 7 if statements (and avoided memcpy) can perhaps explain the performance difference we see. So the performance gain is not due to a single improvement, but a combination of several small gains added together.

File: Objects/listobject.c [hunk 3]
```diff
@@ -1022,21 +1022,28 @@ list_pop_impl(PyListObject *self, Py_ssize_t index)
         PyErr_SetString(PyExc_IndexError, "pop index out of range");
         return NULL;
     }
-    v = self->ob_item[index];
-    if (index == Py_SIZE(self) - 1) {
+
+    PyObject **items = self->ob_item;
+    v = items[index];
+    if(Py_SIZE(self)-1==0) {
```
Comment (corona10) [inline Objects/listobject.c]:
```suggestion
   Py_ssize_t pop_after_size = Py_SIZE(self) - 1;
    if(pop_after_size == 0) {
```

File: Objects/listobject.c [hunk 4]
```diff
@@ -1022,21 +1022,28 @@ list_pop_impl(PyListObject *self, Py_ssize_t index)
...
+        Py_INCREF(v);
+        status = _list_clear(self);
+    }
+    else {
+        if (Py_SIZE(self)-1-index) {
```
Comment (corona10) [inline Objects/listobject.c]:
```suggestion
        if ((pop_after_size - index) > 0) {
```

File: Objects/listobject.c [hunk 5]
```diff
@@ -1022,21 +1022,28 @@ list_pop_impl(PyListObject *self, Py_ssize_t index)
...
+            memmove(&items[index], &items[index+1], (Py_SIZE(self)-1-index)* sizeof(PyObject *));
```
Comment (corona10) [inline Objects/listobject.c]:
```suggestion
            memmove(&items[index], &items[index+1], (pop_after_size - index) * sizeof(PyObject *));
```

File: Objects/listobject.c [hunk 6]
```diff
@@ -1022,21 +1022,28 @@ list_pop_impl(PyListObject *self, Py_ssize_t index)
...
-    Py_INCREF(v);
-    status = list_ass_slice(self, index, index+1, (PyObject *)NULL);
-    if (status < 0) {
-        Py_DECREF(v);
+    if (status >= 0) {
+        return v; // and v now owns the reference the list had 
+    }
+    else {
+        // list resize failed, need to restore
+        memmove(&items[index+1], &items[index], (Py_SIZE(self)-1-index)* sizeof(PyObject *));
```
Comment (corona10) [inline Objects/listobject.c]:
```suggestion
        memmove(&items[index+1], &items[index], (pop_after_size - index)* sizeof(PyObject *));
```

**Review by corona10 — CHANGES_REQUESTED** [review]

File: Objects/listobject.c (hunk 3 above)
Comment (eendebakpt) [reply Objects/listobject.c]: > 

Good suggestion. I renamed to `size_after_pop` instead of the `pop_after_size`.

**Comment by eendebakpt** [comment]
I have made the requested changes; please review again

File: Objects/listobject.c [hunk 7]
```diff
@@ -1022,21 +1022,29 @@ list_pop_impl(PyListObject *self, Py_ssize_t index)
...
-        status = list_resize(self, Py_SIZE(self) - 1);
-        if (status >= 0)
-            return v; /* and v now owns the reference the list had */
-        else
-            return NULL;
+
+    PyObject **items = self->ob_item;
+    v = items[index];
+    const Py_ssize_t size_after_pop = Py_SIZE(self) - 1;
+    if(size_after_pop == 0) {
```
Comment (corona10) [inline Objects/listobject.c]:
```suggestion
    if (size_after_pop == 0) {
```

**Review by corona10 — COMMENTED** [review]
lgtm with nit comment

**Pull request merged**